File: apps/visualizer/backend/utils/lr_catalog_write.py
"""Lightroom catalog write availability for keyword mutations."""

from __future__ import annotations

import os
from dataclasses import dataclass

from lightroom_tagger.core.config import load_config
from lightroom_tagger.lightroom.writer import (
    CULL_KEYWORD,
    KeywordAddResult,
    KeywordRemoveResult,
    add_keyword_by_key,
    backup_catalog_if_needed,
    connect_catalog,
    image_has_keyword_by_key,
    raise_if_catalog_locked,
    remove_keyword_by_key,
)


@dataclass(frozen=True)
class LrCatalogWriteStatus:
    available: bool
    path: str | None
    reason: str | None = None

    def to_dict(self) -> dict:
        return {
            "available": self.available,
            "path": self.path,
            "reason": self.reason,
        }


def describe_lr_catalog_write_status() -> LrCatalogWriteStatus:
    """Report whether the configured Lightroom catalog can accept keyword writes."""
    cfg = load_config()
    path = (cfg.catalog_path or "").strip()
    if not path:
        return LrCatalogWriteStatus(
            available=False,
            path=None,
            reason="No Lightroom catalog configured.",
        )
    if not os.path.isfile(path):
        return LrCatalogWriteStatus(
            available=False,
            path=path,
            reason="Lightroom catalog file not found.",
        )
    try:
        raise_if_catalog_locked(path)
    except RuntimeError as exc:
        return LrCatalogWriteStatus(available=False, path=path, reason=str(exc))
    return LrCatalogWriteStatus(available=True, path=path, reason=None)


def read_cull_keyword_present(image_key: str) -> bool | None:
    """Return whether ``lrt-cull`` is on the image, or None when catalog unavailable."""
    status = describe_lr_catalog_write_status()
    if not status.available or not status.path:
        return None
    conn = connect_catalog(status.path)
    try:
        return image_has_keyword_by_key(conn, image_key, CULL_KEYWORD)
    finally:
        conn.close()


def write_cull_keyword(image_key: str) -> KeywordAddResult:
    status = describe_lr_catalog_write_status()
    if not status.available or not status.path:
        raise RuntimeError(status.reason or "Lightroom catalog unavailable.")
    raise_if_catalog_locked(status.path)
    backup_catalog_if_needed(status.path)
    conn = connect_catalog(status.path)
    try:
        return add_keyword_by_key(conn, image_key, CULL_KEYWORD)
    finally:
        conn.close()


def remove_cull_keyword(image_key: str) -> KeywordRemoveResult:
    status = describe_lr_catalog_write_status()
    if not status.available or not status.path:
        raise RuntimeError(status.reason or "Lightroom catalog unavailable.")
    raise_if_catalog_locked(status.path)
    backup_catalog_if_needed(status.path)
    conn = connect_catalog(status.path)
    try:
        return remove_keyword_by_key(conn, image_key, CULL_KEYWORD)
    finally:
        conn.close()

File: apps/visualizer/backend/src/utils/lr-catalog-write.ts
/**
 * Lightroom catalog write availability for keyword mutations.
 *
 * Availability is re-checked on every call rather than cached: the user can quit or
 * launch Lightroom between two requests, and a stale "available" would mean writing
 * into a catalog Lightroom has open.
 */
import { statSync } from 'node:fs';
import { config, loadLibraryConfig } from '../config.js';
import {
  addKeywordByKey,
  backupCatalogIfNeeded,
  connectCatalog,
  CULL_KEYWORD,
  imageHasKeywordByKey,
  raiseIfCatalogLocked,
  removeKeywordByKey,
  type KeywordAddResult,
  type KeywordRemoveResult,
} from '../lightroom/writer.js';

export interface LrCatalogWriteStatus {
  available: boolean;
  path: string | null;
  reason: string | null;
}

function isFile(path: string): boolean {
  try {
    return statSync(path).isFile();
  } catch {
    return false;
  }
}

/** Whether the configured Lightroom catalog can accept keyword writes, and why not. */
export function describeLrCatalogWriteStatus(): LrCatalogWriteStatus {
  const cfg = loadLibraryConfig(config.LT_CONFIG_YAML);
  const path = (cfg.catalogPath ?? '').trim();
  if (!path) {
    return { available: false, path: null, reason: 'No Lightroom catalog configured.' };
  }
  if (!isFile(path)) {
    return { available: false, path, reason: 'Lightroom catalog file not found.' };
  }
  try {
    raiseIfCatalogLocked(path);
  } catch (e) {
    return { available: false, path, reason: e instanceof Error ? e.message : String(e) };
  }
  return { available: true, path, reason: null };
}

/** Whether `lrt-cull` is on the image, or `null` when the catalog is unavailable. */
export function readCullKeywordPresent(imageKey: string): boolean | null {
  const status = describeLrCatalogWriteStatus();
  if (!status.available || !status.path) return null;
  const conn = connectCatalog(status.path);
  try {
    return imageHasKeywordByKey(conn, imageKey, CULL_KEYWORD);
  } finally {
    conn.close();
  }
}

/**
 * Add `lrt-cull` to the image in the live catalog.
 *
 * Lock is checked again after availability, because Lightroom can open between the
 * two calls.
 */
export function writeCullKeyword(imageKey: string): KeywordAddResult {
  const status = describeLrCatalogWriteStatus();
  if (!status.available || !status.path) {
    throw new RangeError(status.reason ?? 'Lightroom catalog unavailable.');
  }
  raiseIfCatalogLocked(status.path);
  backupCatalogIfNeeded(status.path);
  const conn = connectCatalog(status.path);
  try {
    return addKeywordByKey(conn, imageKey, CULL_KEYWORD);
  } finally {
    conn.close();
  }
}

/** Remove `lrt-cull` from the image in the live catalog. */
export function removeCullKeyword(imageKey: string): KeywordRemoveResult {
  const status = describeLrCatalogWriteStatus();
  if (!status.available || !status.path) {
    throw new RangeError(status.reason ?? 'Lightroom catalog unavailable.');
  }
  raiseIfCatalogLocked(status.path);
  backupCatalogIfNeeded(status.path);
  const conn = connectCatalog(status.path);
  try {
    return removeKeywordByKey(conn, imageKey, CULL_KEYWORD);
  } finally {
    conn.close();
  }
}
